jumanji/environments/routing/macvrp/env.py

```diff
@@ -0,0 +1,614 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Sequence, Tuple
+
+import chex
+import jax
+import matplotlib
+import numpy as np
+from numpy.typing import NDArray
+
+from jumanji import specs
+from jumanji.env import Environment
+from jumanji.environments.routing.macvrp.reward import DenseReward, RewardFn
+from jumanji.environments.routing.macvrp.specs import (
+    NodeSpec,
+    ObservationSpec,
+    PenalityCoeffSpec,
+    TimeWindowSpec,
+    VehicleSpec,
+)
+from jumanji.environments.routing.macvrp.types import (
+    Node,
+    Observation,
+    PenalityCoeff,
+    State,
+    TimeWindow,
+    Vehicle,
+)
+from jumanji.environments.routing.macvrp.utils import (
+    DEPOT_IDX,
+    compute_distance,
+    compute_time_penalties,
+    generate_problem,
+    get_init_settings,
+)
+from jumanji.environments.routing.macvrp.viewer import MACVRPViewer
+from jumanji.types import TimeStep, restart, termination, transition
+from jumanji.viewer import Viewer
+
+
+class MACVRP(Environment[State]):
+    """
+    Multi-Vehicle Routing Problems with Soft Time Windows (MVRPSTW) environment as described in [1].
+    We simplfy the naming to multi-agent capacitated vehicle routing problem (MACVRP).
+
+    - reward: jax array (float32)
+        this global reward is provided to each agent. The reward is equal to the negative sum of the
+        distances between consecutive nodes at the end of the episode over all agents. All time
+        penalties are also added to the reward.
+
+    - observation and state:
+        the observation and state variable types are defined in:
+        jumanji/environments/routing/macvrp/types.py
+
+    [1] Zhang et al. (2020). "Multi-Vehicle Routing Problems with Soft Time Windows: A
+    Multi-Agent Reinforcement Learning Approach".
+    """
+
+    # Use this
+    """Instantiates a `CVRP` environment.
+
+        Args:
+            num_nodes: number of city nodes in the environment. Defaults to 20.
+            max_capacity: maximum capacity of the vehicle. Defaults to 30.
+            max_demand: maximum demand of each node. Defaults to 10.
+            reward_fn: `RewardFn` whose `__call__` method computes the reward of an environment
+                transition. The function must compute the reward based on the current state,
+                the chosen action, the next state and whether the action is valid.
+                Implemented options are [`DenseReward`, `SparseReward`]. Defaults to `DenseReward`.
+            viewer: `Viewer` used for rendering. Defaults to `CVRPViewer` with "human" render mode.
+        """
+
+    def __init__(
+        self,
+        num_customers: int = 6,
+        num_vehicles: int = 2,
```

Comment:
Would it be possible to make a generator that will take these arguments? Like in `Connector` for instance. We are doing this for CVRP as well (cc @surana01).

Reply:
Updated :+1: